compiler/phpdoc.cpp

```diff
@@ -628,3 +616,37 @@ std::vector<std::string> phpdoc_find_tag_as_string_multi(vk::string_view phpdoc,
 bool phpdoc_tag_exists(vk::string_view phpdoc, php_doc_tag::doc_type tag_type) {
   return static_cast<bool>(phpdoc_find_tag_as_string(phpdoc, tag_type));
 }
+
+/*
+ * Sets class_ptr of all nested op_type_expr_class — converting a string class_name to ClassPtr by resolving.
+ * 'class_name' is a relative class name to be resolved on the context of cur_function, it may be "self" or similar also.
+ * We store class_name when parsing phpdoc / type hints, as classes may have not been loaded at the time of parsing.
+ * So, this function must be called for every type rule at some point, where all classes and contexts exist.
+ */
+bool phpdoc_prepare_type_expr_resolving_classes(FunctionPtr cur_function, VertexPtr type_expr) {
+  bool all_resolved = true;
+
+  if (auto as_op_class = type_expr.try_as<op_type_expr_class>()) {
+    if (!as_op_class->class_ptr) {
+      stage::set_location(as_op_class->get_location());
+      const std::string &class_name = resolve_uses(cur_function, as_op_class->class_name, '\\');
+      ClassPtr klass = G->get_class(class_name);
+      as_op_class->class_ptr = klass;
+      if (!klass) {
+        all_resolved = false;
+        kphp_error(0, fmt_format("{}: {}",
+                                 TermStringFormat::paint_red(TermStringFormat::add_text_attribute("Could not find class", TermStringFormat::bold)),
+                                 TermStringFormat::add_text_attribute(class_name, TermStringFormat::underline)));
+      }
+      else {
+        kphp_error(!klass->is_trait(), "You may not use trait as a type-hint");
+      }
+    }
+  }
+
+  for (auto i : *type_expr) {
+    all_resolved &= phpdoc_prepare_type_expr_resolving_classes(cur_function, i);
```

Comment:
nit) It's a bit strange using a `bool` variables with bitwise `&` it would be better to separate these expressions or use `&&` instead

Reply:
Okay, now &&